src/WinRT.Runtime/GuidGenerator.cs

```diff
@@ -342,15 +346,37 @@ internal static Guid CreateIIDForGenericType(string signature)
                 return encode_guid(sha.ComputeHash(data));
             }
 #else
-            var maxBytes = UTF8Encoding.UTF8.GetMaxByteCount(signature.Length);
-
-            var data = new byte[16 /* Number of bytes in a GUID */ + maxBytes];
-            Span<byte> dataSpan = data;
-            wrt_pinterface_namespace.TryWriteBytes(dataSpan);
-            var numBytes = UTF8Encoding.UTF8.GetBytes(signature, dataSpan[16..]);
-            data = data[..(16 + numBytes)];
-
-            return encode_guid(SHA1.HashData(data));
+#nullable enable
+            // Get the maximum UTF8 byte size and allocate a buffer for the encoding.
+            // If the minimum buffer is small enough, we can stack-allocate it.
+            // The 512 threshold is the smallest one that will contain generally all
+            // enum types names. Eg. we'd get signature strings like this:
+            // 'pinterface({61c17706-2d65-11e0-9ae8-d48564015472};enum(Windows.UI.Xaml.Visibility;u4))'.
+            // Which would result in a number of bytes ~280, including the GUID size.
+            int maxUtf8ByteCount = Encoding.UTF8.GetMaxByteCount(signature.Length);
+            int minimumPooledLength = 16 /* Number of bytes in a GUID */ + maxUtf8ByteCount;
+            byte[]? utf8BytesFromPool = null;
+            Span<byte> utf8Bytes = minimumPooledLength <= 512
+                ? stackalloc byte[512]
+                : (utf8BytesFromPool = ArrayPool<byte>.Shared.Rent(minimumPooledLength));
+
+            wrt_pinterface_namespace.TryWriteBytes(utf8Bytes);
+
+            int encodedUtf8BytesWritten = Encoding.UTF8.GetBytes(signature, utf8Bytes[16..]);
+            Span<byte> sha1Bytes = stackalloc byte[SHA1.HashSizeInBytes];
+
+            // Hash the encoded signature (the bytes written are guaranteed to always match the span)
+            _ = SHA1.HashData(utf8Bytes[..(16 + encodedUtf8BytesWritten)], sha1Bytes);
+
+            Guid iid = encode_guid(sha1Bytes);
+
+            // Before exiting, make sure to return the array from the pool, if we rented one
+            if (utf8BytesFromPool is not null)
+            {
+                ArrayPool<byte>.Shared.Return(utf8BytesFromPool);
+            }
+
+            return iid;
```

Comment:
Also you want to restore the nullable scope at the end here, otherwise we will run into issues in the scenario / repo that merges these files.